src/ui.rs

```diff
@@ -56,18 +56,35 @@ pub fn draw<B: Backend>(f: &mut Frame<B>, app: &mut App) {
 
 fn draw_chat<B: Backend>(f: &mut Frame<B>, app: &mut App, area: Rect) {
     let text_width = area.width.saturating_sub(2) as usize;
-    let lines = app
-        .data
-        .input
-        .chars()
-        .enumerate()
-        .fold(Vec::new(), |mut lines, (idx, c)| {
-            if idx % text_width == 0 {
-                lines.push(String::new())
-            }
-            lines.last_mut().unwrap().push(c);
-            lines
-        });
+    let lines: Vec<String> =
+        app.data
+            .input
+            .chars()
+            .enumerate()
+            .fold(Vec::new(), |mut lines, (idx, c)| {
+                if idx % text_width == 0 {
+                    lines.push(String::new());
+                }
+                match c {
+                    '\n' => {
+                        lines.last_mut().unwrap().push('\n');
+                        lines.push(String::new())
+                    }
+                    _ => lines.last_mut().unwrap().push(c),
+                }
+                lines
+            });
+    // chars since newline on `cursor_y` line
```

Comment:
@goolord: Ah, I see. The input handling is a little bit hacky. I think I have a solution though. You can calculate here:

```rust
let last_line_width = lines.last().map(|l| l.width()).unwrap_or(0);
let num_input_lines = lines.len().max(1);
```

Then add `last_line_width + 1` to `x`, and `num_input_lines` to `y` in lines 114, 116.

Reply:
the cursor isn't always on the last line, so this code calculates which line its on and the offset from the beginning of that line